pyscriptjs/src/plugins/pyterminal.ts

```diff
@@ -89,38 +130,117 @@ function make_PyTerminal(app: PyScriptApp) {
             this.outElem.className = 'py-terminal';
             this.appendChild(this.outElem);
 
-            if (this.isAuto()) {
-                this.classList.add('py-terminal-hidden');
-                this.autoShowOnNextLine = true;
-            } else {
-                this.autoShowOnNextLine = false;
-            }
+            this.setupPosition(app);
+        }
 
+        // implementation of the Stdio interface
+        stdout_writeline(msg: string) {
+            this.outElem.innerText += msg + '\n';
             if (this.isDocked()) {
-                this.classList.add('py-terminal-docked');
+                this.scrollTop = this.scrollHeight;
+            }
+            if (this.autoShowOnNextLine) {
+                this.classList.remove('py-terminal-hidden');
+                this.autoShowOnNextLine = false;
             }
+        }
 
-            logger.info('Registering stdio listener');
-            app.registerStdioListener(this);
+        stderr_writeline(msg: string) {
+            this.stdout_writeline(msg);
         }
+        // end of the Stdio interface
+    }
 
-        isAuto() {
-            return this.hasAttribute('auto');
+    return PyTerminalPre;
+}
+
+declare const Terminal: typeof TerminalType;
+
+function make_PyTerminal_xterm(app: PyScriptApp) {
+    /** The <py-terminal> custom element, which automatically register a stdio
+     *  listener to capture and display stdout/stderr
+     */
+    class PyTerminalXterm extends PyTerminalBaseClass {
+        outElem: HTMLDivElement;
+        moduleResolved: boolean;
+        xtermReadyPromise: Promise<TerminalType>;
+        xterm: TerminalType;
+        cachedStdOut: Array<string>;
+        cachedStdErr: Array<string>;
+
+        constructor() {
+            super();
+            this.moduleResolved = false;
+            this.cachedStdOut = [];
+            this.cachedStdErr = [];
+
+            //Required to make xterm appear properly
+            this.style.width = '100%';
+            this.style.height = '100%';
         }
 
-        isDocked() {
-            return this.hasAttribute('docked');
+        async connectedCallback() {
+            this.outElem = document.createElement('div');
+            //this.outElem.className = 'py-terminal';
+            this.appendChild(this.outElem);
+
+            this.setupPosition(app);
+
+            this.xtermReadyPromise = this.setupXterm();
+            await this.xtermReadyPromise;
+        }
+
+        async setupXterm() {
+            // eslint-disable-next-line
+            // @ts-ignore
+            if (globalThis.Terminal == undefined) {
+                //xterm module proper
+                //eslint-disable-next-line
+                //@ts-ignore
+                await import('https://cdn.jsdelivr.net/npm/xterm@5.1.0/lib/xterm.js');
```

Comment:
we have `xterm` as dependency of this module but here we're awaiting an hard-coded *xterm* version that might not be the same we use as dependency ... what is the need for this CDN request, since we bundle code and bundlers can (usually) deal with dynamic imports avoiding the need for a network request at all?

Reply:
This may be me not fully understanding how dynamic imports work - my intention was to only load the xterm.js package over the wire if necessary, i.e. if we're using _xterm_ py-terminals. My thinking was, if we can spare (the majority of?) users the ~275kb network request, we should. It might seem like minimal gains, but since we're optimizing toward a ~2.2 MB total size _including the interpreter_ with Micropython, including xterm.js in every build would then be a non-trivial percentage of the total files fetched, for a feature I think most users aren't likely to use.

Hopefully that thinking is clear... and if there's a better way to do this I'm more than game.

Reply:
the `_xterm_cdn_base_url` moved in a single place would be already an answer and I didn't know *xterm* was 275K so I am with you in here in making its evaluation lazy 👍